src/components/input/input.jsx

```diff
@@ -1,69 +1,104 @@
 import PropTypes from 'prop-types';
-import React from 'react';
+import React, {
+  forwardRef,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+} from 'react';
 import cautionSvg from '../../svgs/caution.svg';
 import FormControl from '../form-control/form-control.jsx';
 import Icon from '../icon/icon.jsx';
 import styles from './input.css';
+import useDidMount from '../../hooks/use-did-mount.js';
+import useValidation from '../../hooks/use-validation.jsx';
 
-function getClassName(className, invalid, readOnly) {
+function getClassName(className, validationMessage) {
   if (className) return className;
-  return isInvalid(invalid, readOnly) ? styles['invalid-input'] : styles.input;
+  return validationMessage ? styles['invalid-input'] : styles.input;
 }
 
-function isInvalid(invalid, readOnly) {
-  return invalid && !readOnly;
-}
+const Input = forwardRef(function Input(props, forwardedRef) {
+  const fallbackRef = useRef();
+  const ref = forwardedRef || fallbackRef;
+  const inputRef = useRef();
+  const [didMount] = useDidMount();
+  const [validationMessage, validate] = useValidation(props.validationMessage, inputRef);
+
+  function onBlur(event) {
+    validate();
+    props.onBlur(event);
+  }
+
+  function onChange(event) {
+    validate();
+    props.onChange(event);
+  }
+
+  useEffect(() => {
+    if (!didMount) return;
+
+    // The MDS Input is using an uncontrolled `<input>`.
+    // In order to set a new provided value prop, we
+    // set the internal state of the `<input>`.
+    inputRef.current.value = props.value;
+  }, [props.value]);
+
+  useImperativeHandle(ref, () => ({
+    get dirty() {
+      const providedValue = props.value || '';
+      return providedValue !== this.value;
+    },
+    id: props.id,
+    name: props.name,
+    get value() {
+      return inputRef.current.value;
+    },
+  }));
 
-export default function Input(props) {
   return (
     <FormControl
       className={props.cssContainer}
-      error={props.invalid ? 'Invalid value.' : ''}
+      error={validationMessage}
       id={props.id}
       label={props.label}
       readOnly={props.readOnly}
       required={props.required}
     >
       <input
+        aria-describedby={`${props.id}Hint`}
         autoFocus={props.autoFocus} // eslint-disable-line jsx-a11y/no-autofocus
-        className={getClassName(props.className, props.invalid, props.readOnly)}
-        defaultValue={props.defaultValue}
-        disabled={props.disabled}
+        className={getClassName(props.className, validationMessage)}
+        defaultValue={props.value}
```

Comment:
We are no longer supporting a controlled input. We are always rendering an uncontrolled input -- but we update the internal `<input>` state whenever a new `value` prop is given. 

If the developer wants to hoist input state then they can listen to events -- either `onChange` or handle bubbled events. Same for all other events. If a developer want to avoid hoisting then they can use the `ref` API on the Input.